File: twist.c
/* File twist.c
 * created/modified on Son Nov  8 17:59:09 CET 2015

 * by creac for Son Nov  8 17:59:09 CET 2015

 */

#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <string.h>
#include <assert.h>
#include <stdlib.h>

#  define STRERR1  "Error in the arguments"
#  define STRERR2  "Can't open input file(s)"
#  define STRERR3  "Can't stat input file(s)"
#  define STRERR4  "Can't open output file"
#  define STRERR5  "Synch error -- empty file?"
#  define FORMAT   "USAGE: "
#  define STRWARN1 "Warning: files have not the same size"
#  define STRMSG1  "Processing starts..."
#  define STRMSG2  "Processing ends."
#  define STRMSG3  "Remember"

#define ERR1	-1
#define MAXTWEEPS 500
#define SIZEOFPICTURE    24	// amount of characters corresponding to a picture

char verbose;

void usage(void);

int main (int argc, char *argv[])
{
	int i, l, sum, tweet_length;
	char *tweet;
	char *tweeps_file; FILE *tf;
	char *out_file = NULL;
	char dontTweet = 0;
	char first_time=1;
	int ntweeps = MAXTWEEPS;
	char **tweeps, **tweep;
	char * line = NULL;
	size_t len = 0;
	ssize_t read;
	int n;
	char *p;
	char buffer[141];
	char *picture_file = NULL;
	char format_string[64];

	if (argc == 1) {
		usage();
		return 0;
	}
	for (i=1; i<argc; i++) {
		if (argv[i][0] == '-')
		switch(argv[i][1]) {
		case 'm':  // message
			tweet = strdup(argv[++i]);
			break;
		case 'l':  // tweep list
			tweeps_file = argv[++i];
			break;
		case 's':  // simulate (do not tweet)
			dontTweet = 1;
			break;
		case 'v':  // verbose
			verbose = 1;
			break;
		case '#':  // max no. of tweeps
			sscanf(argv[++i], "%d", &ntweeps);
			assert(ntweeps > 0);
			break;
		case 'o': // output file
			out_file = argv[++i];
			break;
		case 'p': // picture file
			picture_file = argv[++i];
			break;
		default: usage(); return ERR1;
		}
	}

	// file system operations
	tf = fopen(tweeps_file, "r");
	assert (tf != NULL);

	tweet_length = strlen(tweet);
	assert (tweet_length < 140);

	tweeps = (char **) malloc(ntweeps * sizeof(char *));
	
	for (n=0; ((read = getline(&line, &len, tf)) != -1); n++) {
		p = tweeps[n] = strdup(line);
		p = strchr(p, '\n');
		assert(p != NULL);
		*p = ' '; // THIS IS A SPACE, USED TO SEPARATE TWEEPS ON THE OUTPUT TWEET
		if (verbose) printf("Tweep[%d] = %s\n", n, tweeps[n]);
	}

	fclose(tf);
	if (line) free(line);

	tf = fopen(out_file, "w");
	assert (tf != NULL);

	// constructing the tweets
	tweep = tweeps;
	i=0;
	first_time=1;

	while (i<n) {
		*buffer = '\0';
		// sum = (picture_file == NULL)? 0:SIZEOFPICTURE;
		// strcpy(buffer, tweet);
		// strcat(buffer, " ");
		for (sum=tweet_length+1;    i<n;    i++, tweep++) {

			l = strlen(*tweep);
			// NOTA: COSI' FUNZIONA if (sum+l>137) break;
			if (sum+l>138) break;
			strcat(buffer, *tweep);
			//sum = strlen(buffer);
			if (verbose) printf("buffer=%s\n", buffer);
			sum += l;
		}

		// only single space chars please
		do { char *s;
			s = strdup(buffer);
			p=strstr(s, "  ");
			if (p==NULL) break;
			*p = '\0';
			sprintf(buffer, "%s%s", s, p+1);
			free(s);
		} while (1);

		len = strlen(buffer)+tweet_length+1;
		assert(len<=139);

		if (first_time)
		{
		printf("TWEET: %s %s (nchar=%zd)\n", tweet, buffer, len), first_time=0;
		//sprintf(format_string, "sudo t update '%%s %%s' -f %s '%%s %%s'", picture_file
		fprintf(tf, "sudo t update '%s %s'\n", tweet, buffer);
		}
		else
		{
		printf("TWEET: %s%s (nchar=%zd)\n", buffer, tweet, len);
		fprintf(tf, "sudo t update '%s%s'\n", buffer, tweet);
		}
	}

	// closings
	fclose(tf);
	for (i=0; i<n; i++) free(tweeps[i]); 
	free(tweeps);
}

void usage(void)
{
fprintf(stderr, "twist: %s\n", STRERR1);
fprintf(stderr, "usage: twist -m messageToTweet -l tweepList -s -v\n");
fprintf(stderr, "       -# maxNumberOfTweeps -o outputScript\n");
//fprintf(stderr, "       -# maxNumberOfTweeps -o outputScript -p pictureFilename\n");
fprintf(stderr, "       -s: simulate (do not tweet), -v: verbose\n");
}
/* End of file twist.c */

File: twist.go
/* File twist.go
 * created by Vincenzo De Florio on Die Apr 12 12:18:52 CEST 2016
 * translated from twist.c, by the same author
 */

package	main
import	(
	"fmt"
	"os"
	"log"
	"strconv"
	"io"
)

func Strchr(s []byte, delim byte) *byte {
	var i int
	for i=0; i<len(s); i++ {
		if s[i] == byte {
			return &(s[i])
		}
	}
	return nil
}

func check(e error) {
    if e != nil {
        panic(e)
    }
}

const ERR1 = -1
const MAXTWEEPS = 500
const SIZEOFPICTURE = 24


func assert (condition bool) () {
	if condition == false {
		log.Fatalln("assertion failed") // Fatal(err)
		// Fatalln calls os.Exit(1)
	}
}


func main () {
	// #CLANG argc argv
	argv := os.Args
	argc := len(argv)

	Verbose := false
	var err error
	var i, l, sum, tweet_length int
	var tweet, tweeps_file string
	var tf *File
	var out_file string = nil
	// char *out_file = NULL
	var dontTweet bool
	// char dontTweet = 0
	//char first_time=1
	var first_time bool = false
	//int ntweeps = MAXTWEEPS
	ntweeps := MAXTWEEPS
	var tweeps, tweep *string
	// char **tweeps, **tweep
	//char * line = NULL
	var line string = nil
	len := 0
	// size_t len = 0
	// ssize_t read
	var read int
	// int n
	var n int
	// char *p
	var p *byte
	var buffer string // char buffer[141]
	// char *picture_file = NULL
	var picture_file string = nil
	//char format_string[64]
	var format_string string

	if argc == 1 {
		usage()
		return
	}
	for i=1; i<argc; i++ {
		if argv[i][0] == '-' {
			switch argv[i][1] {
			case 'm':  // message
				// instead of tweet = strdup(argv[++i]), I just do...
				i++
				tweet = argv[i]	// note the ":=" operator

			case 'l':  // tweep list
				i++
				tweeps_file = argv[i]
			case 's':  // simulate (do not tweet)
				dontTweet = 1
			case 'v':  // verbose
				Verbose = 1
			case '#':  // max no. of tweeps
				i++
				// #CLANG sscanf(argv[i], "%d", &ntweeps)
				// ntweeps = strconv.ParseInt(argv[i],10,32)
// i, err := strconv.Atoi("-42")
// s := strconv.Itoa(-42)
				ntweeps, err = strconv.Atoi(argv[i])
				if err != nil {
					usage()
					os.Exit(ERR1)
				}
				assert (ntweeps > 0)
			case 'o': // output file
				i++
				out_file = argv[i]
			case 'p': // picture file
				i++
				picture_file = argv[i]
			default:
				usage()
				os.Exit(ERR1)
			}
		}
	}

	// file system operations
	// tf = fopen(tweeps_file, "r")
	tf, err = os.Open(tweeps_file)
	if (err != nil) {
		log.Fatal(err)
	}
	// assert (err != NULL)

	tweet_length = len(tweet)
	assert (tweet_length < 140)

	// tweeps = (char **) malloc(ntweeps * sizeof(char *))
	tweeps = make([]string, ntweeps)
	
	for n=0; ; n++ {
		line, err = ReadString('\n')
		if err == io.EOF {
			break
		}
		tweeps[n] = line
		p = &tweeps[n]
		p = Strchr(p, '\n')
		assert(p != nil)
		*p = ' '; // THIS IS A SPACE, USED TO SEPARATE TWEEPS ON THE OUTPUT TWEET
		if Verbose {
			fmt.Printf("Tweep[%d] = %s\n", n, tweeps[n])
		}
	}

	tf.Close()
	assert (err != nil)
	// if line free(line)  NO NEED FOR THIS

	tf, err = os.Create(out_file)
	check(err)

	// constructing the tweets
	tweep = tweeps
	first_time=true

	for i=0; i<n;  {
		buffer = nil
		// sum = (picture_file == NULL)? 0:SIZEOFPICTURE
		// strcpy(buffer, tweet)
		// strcat(buffer, " ")
		for sum=tweet_length+1;    i<n;    i=i+1 {
			l = len(tweep[i])
			// NOTA: COSI' FUNZIONA if sum+l>137 break
			if sum+l>138 { break }
			strcat(buffer, tweep[i])
			//sum = strlen(buffer)
			if verbose { fmt.Printf("buffer=%s\n", buffer) }
			sum += l
			//++tweep
		}

		// only single space chars please
		for ;; {
			//s = buffer
			//p=strstr(s, "  ")
			// func Index(s, sep string) int
			// Index returns the index of the first instance of sep in s, or -1 if sep is not present in s.
			idx := Index(s, "  ")
			if idx == -1 { break }
			buffer = buffer[0:idx] + buffer[idx+1:]
		}

		len = len(buffer)+tweet_length+1
		assert(len<=139)

		if first_time==true {
			fmt.Printf("TWEET: %s %s (nchar=%zd)\n", tweet, buffer, len)
			first_time=false
			//sprintf(format_string, "sudo t update '%%s %%s' -f %s '%%s %%s'", picture_file
			fmt.Fprintf(tf, "sudo t update '%s %s'\n", tweet, buffer)
		} else {
			fmt.Printf("TWEET: %s%s (nchar=%zd)\n", buffer, tweet, len)
			fmt.Fprintf(tf, "sudo t update '%s%s'\n", buffer, tweet)
		}
	}

	// closings
	tf.Close()
	//for i=0; i<n; i++ free(tweeps[i]); 
	//free(tweeps)
}

func usage() {
	fmt.Fprintln(os.Stderr,  "twist: Error in the calling arguments\n")
	fmt.Fprintln(os.Stderr,  "usage: twist -m messageToTweet -l tweepList -s -v\n")
	fmt.Fprintln(os.Stderr,  "       -# maxNumberOfTweeps -o outputScript\n")
	//fmt.Fprintln(os.Stderr,  "       -# maxNumberOfTweeps -o outputScript -p pictureFilename\n")
	fmt.Fprintln(os.Stderr,  "       -s: simulate (do not tweet), -v: verbose\n")
}
/* End of file twist.go */
